plugins/parsers/opentsdb/parser.go

```diff
@@ -0,0 +1,118 @@
+package opentsdb
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+// Parser encapsulates a OpenTSDB Parser.
+type Parser struct {
+	DefaultTags map[string]string `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var metrics []telegraf.Metric
+
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// delete LF and CR
+		line = strings.TrimRight(line, "\r\n")
+
+		m, err := p.ParseLine(line)
+		if err != nil {
+			return nil, err
+		}
+
+		if m != nil {
+			metrics = append(metrics, m)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return metrics, nil
+}
+
+// ParseLine performs OpenTSDB parsing of a single line.
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	// Break into fields ("put", name, timestamp, value, tag1, tag2, ..., tagN).
+	fields := strings.Fields(line)
+	if len(fields) == 0 {
+		return nil, nil
+	}
+	if len(fields) < 4 || fields[0] != "put" {
+		return nil, fmt.Errorf("received %q which doesn't have required fields", line)
+	}
+
+	// decode the name and tags
+	measurement := fields[1]
+	tsStr := fields[2]
+	valueStr := fields[3]
+	tagStrs := fields[4:]
+
+	// Parse value.
+	v, err := strconv.ParseFloat(valueStr, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing field %q value failed: %w", measurement, err)
+	}
+
+	fieldValues := map[string]interface{}{"value": v}
+
+	// Parse timestamp.
+	ts, err := strconv.ParseInt(tsStr, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing field %q time failed: %w", measurement, err)
+	}
+
+	var timestamp time.Time
+	if ts < 1e12 {
+		// second resolution
+		timestamp = time.Unix(ts, 0)
+	} else {
+		// millisecond resolution
+		timestamp = time.UnixMilli(ts)
+	}
+
+	tags := make(map[string]string, len(p.DefaultTags)+len(tagStrs))
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	for _, tag := range tagStrs {
+		tagValue := strings.Split(tag, "=")
+		if len(tagValue) != 2 {
+			continue
+		}
+
+		name := tagValue[0]
+		value := tagValue[1]
+		if name == "" || value == "" {
+			continue
+		}
```

Comment:
I don't see in the specs that a tag with empty string as value would not be accepted?

Reply:
Hmm, I see:

```
put <metric> <timestamp> <value> <tagk_1>=<tagv_1>[ <tagk_n>=<tagv_n>]
```

The use of `<tagv_1>` and `<tagv_n>` says to me that they are mandatory. If they were optional, it would be `[tagv_1]` and `[tagv_n]`, wouldn't it?

Reply:
There are no length constraints defined. Only that it can't have spaces..

Reply:
Okay, fair enough. Just wanted to make sure telegraf wasn't more strict than OpenTSDB itself 😉